### PR TITLE
Sailjail validation

### PR DESCRIPTION
**Pull request by vigejolla**

### CODE REVIEW

File: rpmvalidation.sh [hunk 1]
```diff
@@ -539,6 +550,52 @@ validatedesktopfile() {
...
+    local value=$(echo $1 | sed "s/[^=]*=\(.*\)/\1/")
+    if ! check_contained_in "$key" $ALLOWED_SAILJAILKEYS; then
+        validation_error $DESKTOP_NAME "X-Sailjail key is not allowed: $key"
+        INFO_MSG_PRINTED=1
+    elif [[ $key == Permissions ]] ; then
+        echo "$value" | validatesailjailpermissions
+    elif [[ $key == OrganizationName ]]; then
+        validateorganizationname "$value"
+    elif [[ $key == ApplicationName ]]; then
+        [[ $value =~ [.] ]] && validation_error "ApplicationName may not contain '.'"
```
Comment (Tomin1) [inline rpmvalidation.sh:570]: There should be a similar character validation that organization name has. E.g. `[[ ! $value =~ ^[a-z-_][0-9a-z-_]*$ ]]`.

File: rpmvalidation.sh [hunk 2]
```diff
@@ -539,6 +550,52 @@ validatedesktopfile() {
...
+    fi
+}
+
+validateorganizationname () {
+    if [[ ! $1 =~ ^[0-9a-z.-_]+$ ]]; then
+        validation_error $DESKTOP_NAME "Organization name contains illegal characters"
+        INFO_MSG_PRINTED=1
+    fi
+    if [[ $1 =~ [.][0-9] ]]; then
```
Comment (Tomin1) [inline rpmvalidation.sh:579]: This checks that later parts don't begin with a number but it doesn't check the very first part.

File: rpm/sdk-harbour-rpmvalidator.spec [hunk 3]
```diff
@@ -85,6 +85,9 @@ install -D -m 0644 deprecated_libraries.conf %{buildroot}%{_datadir}/%{name}/dep
 install -D -m 0644 deprecated_qmlimports.conf %{buildroot}%{_datadir}/%{name}/deprecated_qmlimports.conf
 install -D -m 0644 deprecated_requires.conf %{buildroot}%{_datadir}/%{name}/deprecated_requires.conf
 install -D -m 0644 disallowed_qmlimport_patterns.conf %{buildroot}%{_datadir}/%{name}/disallowed_qmlimport_patterns.conf
+install -D -m 0644 allowed_sailjailkeys.conf %{buildroot}%{_datadir}/%{name}/allowed_sailjailkeys.conf
+install -D -m 0644 allowed_permissions.conf %{buildroot}%{_datadir}/%{name}/allowed_permissions.conf
+install -D -m 0644 disallowed_orgnames.conf %{buildroot}%{_datadir}/%{name}/disallowed_orgnames.conf
```
Comment (Tomin1) [inline rpm/sdk-harbour-rpmvalidator.spec:90]: I don't see such file here.

**Review by Tomin1 — CHANGES_REQUESTED** [review]

File: rpmvalidation.sh (hunk 1 above)
Comment (vigejolla) [reply rpmvalidation.sh:570]: The freedesktop spec allows Capital letters in ApplicationName. Should we be more strict in that respect?

Comment (Tomin1) [reply rpmvalidation.sh:570]: No, we should allow capital letters IMO.

Comment (vigejolla) [reply rpmvalidation.sh]: fixed

File: rpmvalidation.sh (hunk 2 above)
Comment (vigejolla) [reply rpmvalidation.sh]: fixed

File: rpm/sdk-harbour-rpmvalidator.spec (hunk 3 above)
Comment (vigejolla) [reply rpm/sdk-harbour-rpmvalidator.spec:90]: forgot to add file... added

**Comment by Tomin1** [comment]
I saw some strange things:
If I have an empty X-Sailjail section without anything there, I get this:

> ERROR [/usr/share/applications/mediagallery.desktop] X-Sailjail key is not allowed: 

Seems a little confusing, otherwise doesn't matter very much. An empty section would likely leave the app pretty broken. It's pretty much equivalent to saying that you don't need any directories, D-Bus names (e.g. coveraction) or any permissions but we haven't really thought out if anyone should leave the section empty. We actually use empty Permissions= key ourselves but leaving it out should work the same way.

If I have a permission that is not allowed, I get this:

> Desktop file
> ============
> ERROR [/usr/share/applications/mediagallery.desktop] X-Sailjail permission not allowed: UDisks
> PASSED

So there is an error but it's also PASSED instead of FAILED? The other new error cases result in FAILED.

**Review by Tomin1 — APPROVED** [review]

**Comment by vigejolla** [comment]
> If I have an empty X-Sailjail section without anything there, I get this:
> 
> ```
> ERROR [/usr/share/applications/mediagallery.desktop] X-Sailjail key is not allowed: 
> ```
> 
> Seems a little confusing, otherwise doesn't matter very much. An empty section would likely leave the app pretty broken. It's pretty much equivalent to saying that you don't need any directories, D-Bus names (e.g. coveraction) or any permissions but we haven't really thought out if anyone should leave the section empty. We actually use empty Permissions= key ourselves but leaving it out should work the same way.

I think we should not allow empty X-Sailjail section, at the minimum it should be:

    [X-Sailjail]
    Permissions=

I'll fix the error message

> Desktop file
> ============
> ERROR [/usr/share/applications/mediagallery.desktop] X-Sailjail permission not allowed: UDisks
> PASSED

I thought I fixed this already, but I guess not then.

**Comment by vigejolla** [comment]
> > # Desktop file
> > ERROR [/usr/share/applications/mediagallery.desktop] X-Sailjail permission not allowed: UDisks
> > PASSED
> 
> I thought I fixed this already, but I guess not then.

Yes, I indeed had fixed this already. Are you sure you were testing with the latest version?

**Review by Tomin1 — APPROVED** [review]
LGTM.

**Comment by HengYeDev** [comment]
Does this mean that the program can write to config/OrganizationName/AppName

**Comment by vigejolla** [comment]
> Does this mean that the program can write to config/OrganizationName/AppName

Sorry for the late reply. That's indeed the idea.

**Pull request merged**